prometheus/collector.go

```diff
@@ -0,0 +1,172 @@
+		"Offset lag of a topic/partition",
+		[]string{"group_id", "consumer_address", "client_id", "topic", "partition"},
+		nil)
+)
+
+// PartitionInfoCollector is a Kafka prometheus.Collector. It uses a
+// exporter.ConsumerGroupInfoClient for the actual querying of Kafka.
+//
+// To speed up collection each consumer group is collected
+// concurrently.
+type PartitionInfoCollector struct {
+	groupListErrors     prometheus.Counter
+	groupDescribeErrors *prometheus.CounterVec
+
+	client exporter.ConsumerGroupInfoClient
+
+	ctx                  context.Context
+	execTimeout          time.Duration
+	maxConcurrentQueries int
+}
+
+// NewPartitionInfoCollector returns a prometheus.Collector that queries Kafka
+// using client. concurrency sets an upper limit on the number concurrent Kafka
+// concumer group queries running.
+func NewPartitionInfoCollector(client exporter.ConsumerGroupInfoClient, ctx context.Context, execTimeout time.Duration, maxConcurrentQueries int) *PartitionInfoCollector {
+	if maxConcurrentQueries <= 0 {
+		log.Fatal("maxConcurrentQueries must be positive.")
+	}
+	return &PartitionInfoCollector{
+		prometheus.NewCounter(prometheus.CounterOpts{
+			Name: "kafka_broker_consumer_group_list_errors",
+			Help: "Number of Kafka scraping errors.",
+		}),
+		prometheus.NewCounterVec(prometheus.CounterOpts{
+			Name: "kafka_broker_consumer_group_describe_errors",
+			Help: "Number of Kafka scraping errors.",
+		}, []string{"group"}),
+		client,
+		ctx,
+		execTimeout,
+		maxConcurrentQueries,
+	}
+}
+
+// Describe transmits all metric descriptions to c.
+func (p *PartitionInfoCollector) Describe(c chan<- *prometheus.Desc) {
+	c <- partitionOffsetMetricsDesc
+	c <- partitionLagMetricsDesc
+	p.groupListErrors.Describe(c)
+	p.groupDescribeErrors.Describe(c)
+}
+
+func (p *PartitionInfoCollector) collectGroupsFromChan(<-chan string) {
+
+}
+
+// Collect triggers an on-demand scraping from Kafka and transmits metrics into
+// c.
+func (p *PartitionInfoCollector) Collect(c chan<- prometheus.Metric) {
+	// Important that these are collected _after_ the Kafka collection below to
+	// correctly accomodate for the errors that happened during the scrape.
+	defer p.groupDescribeErrors.Collect(c)
+	defer p.groupListErrors.Collect(c)
+
+	ctx, cancel := context.WithTimeout(p.ctx, p.execTimeout)
+	groupnames, err := p.client.Groups(ctx)
+	cancel()
+	if err != nil {
+		log.Error("Could not list groups:", err)
+		p.groupListErrors.Inc()
+		return
+	}
+
+	var wg sync.WaitGroup
+	wg.Add(p.maxConcurrentQueries)
+	groupsToProcess := make(chan string)
+	collectGroupWorker := func() {
+		defer wg.Done()
+		for groupname := range groupsToProcess {
+			ctx, cancel := context.WithTimeout(p.ctx, p.execTimeout)
+			partitions, err := p.client.DescribeGroup(ctx, groupname)
+			cancel()
+			if err != nil {
+				log.Error("Could not describe group '%s': %s", groupname, err)
+				p.groupDescribeErrors.WithLabelValues(groupname).Inc()
+				return
```

Comment:
shouldn't this be changed to `continue` to keep processing remaining groupnames?